votor/src/consensus_rewards.rs

```diff
@@ -0,0 +1,275 @@
+                    match msg {
+                        Ok(msg) => {
+                            let (skip, notar) = self.build_certs(msg.slot);
+                            let resp = BuildRewardCertsResponse {
+                                skip,
+                                notar,
+                            };
+                            if self.reward_certs_sender.send(resp).is_err() {
+                                warn!("cert sender channel is disconnected; exiting.");
+                                break;
+                            }
+                        }
+                        Err(_) => {
+                            warn!("build reward certs channel is disconnected; exiting.");
+                            break;
+                        }
+                    }
+                }
+                recv(self.votes_receiver) -> msg => {
+                    match msg {
+                        Ok(msg) => {
+                            for entry in msg.votes {
+                                self.add_vote(msg.root_slot, entry.max_validators, &entry.vote);
+                            }
+                        }
+                        Err(_) => {
+                            warn!("votes receiver channel is disconnected; exiting.");
+                            break;
+                        }
+                    }
+                }
+                default(Duration::from_secs(1)) => {
+                    continue;
+                }
+            }
+        }
+    }
+
+    /// Returns [`true`] if the rewards container is interested in this vote else [`false`].
+    fn wants_vote(&self, root_slot: Slot, vote: &VoteMessage) -> bool {
+        if !wants_vote(&self.cluster_info, &self.leader_schedule, root_slot, vote) {
+            return false;
+        }
+        let Some(entry) = self.votes.get(&vote.vote.slot()) else {
+            return true;
+        };
+        entry.wants_vote(vote)
+    }
+
+    /// Adds received [`VoteMessage`]s from other validators.
+    fn add_vote(&mut self, root_slot: Slot, max_validators: usize, vote: &VoteMessage) {
+        // drop old state no longer needed
+        self.votes = self.votes.split_off(
+            &(root_slot
+                .saturating_add(NUM_SLOTS_FOR_REWARD)
+                .saturating_add(1)),
+        );
+
+        if !self.wants_vote(root_slot, vote) {
+            return;
+        }
+        match self
+            .votes
+            .entry(vote.vote.slot())
+            .or_insert(Entry::new(max_validators))
+            .add_vote(vote)
+        {
+            Ok(()) => (),
+            Err(e) => {
+                warn!("Adding vote {vote:?} failed with {e}");
+            }
+        }
+    }
+
+    /// Builds reward certificates.
+    fn build_certs(
+        &self,
+        slot: Slot,
+    ) -> (
+        Option<SkipRewardCertificate>,
+        Option<NotarRewardCertificate>,
+    ) {
+        match self.votes.get(&slot) {
```

Comment:
if you like the suggestion from here https://github.com/anza-xyz/alpenglow/pull/639/changes#r2639468328

this could remove the entry, and consume it (which consumes the partial certs) to produce a `BuildRewardsCertResponse` instead